include/oneapi/dpl/pstl/execution_impl.h

```diff
@@ -105,25 +78,135 @@ template <typename _ExecutionPolicy>
 using __allow_parallel = typename __internal::__policy_traits<::std::decay_t<_ExecutionPolicy>>::__allow_parallel;
 
 template <typename _ExecutionPolicy, typename... _IteratorTypes>
-auto
-__is_vectorization_preferred(_ExecutionPolicy& __exec)
-    -> decltype(__internal::__lazy_and(__exec.__allow_vector(),
-                                       __internal::__is_random_access_iterator_t<_IteratorTypes...>()))
+constexpr auto
+__is_vectorization_preferred()
 {
-    return __internal::__lazy_and(__exec.__allow_vector(),
-                                  __internal::__is_random_access_iterator_t<_IteratorTypes...>());
+    return ::std::conjunction<decltype(::std::decay_t<_ExecutionPolicy>::__allow_vector()),
+                              __internal::__is_random_access_iterator_t<_IteratorTypes...>>();
 }
 
 template <typename _ExecutionPolicy, typename... _IteratorTypes>
-auto
-__is_parallelization_preferred(_ExecutionPolicy& __exec)
-    -> decltype(__internal::__lazy_and(__exec.__allow_parallel(),
-                                       __internal::__is_random_access_iterator_t<_IteratorTypes...>()))
+constexpr auto
+__is_parallelization_preferred()
+{
+    return ::std::conjunction<decltype(::std::decay_t<_ExecutionPolicy>::__allow_parallel()),
+                              __internal::__is_random_access_iterator_t<_IteratorTypes...>>();
+}
+
+//------------------------------------------------------------------------
+// backend selector with tags
+//------------------------------------------------------------------------
+
+#if _ONEDPL_PAR_BACKEND_TBB
+using __par_backend_tag = __tbb_backend_tag;
+#elif _ONEDPL_PAR_BACKEND_OPENMP
+using __par_backend_tag = __omp_backend_tag;
+#elif _ONEDPL_PAR_BACKEND_SERIAL
+using __par_backend_tag = __serial_backend_tag;
+#else
+#    error "Parallel backend was not specified"
+#endif
+
+template <class _IsVector>
+struct __serial_tag
+{
+    using __is_vector = _IsVector;
+};
+
+template <class _IsVector>
+struct __parallel_tag
```

Comment:
I would propose to add a parameter `_BackendType` to this template definition:

```
template <typename _IsVector, typename _BackendType = __par_backend_tag>
struct __parallel_tag
{
    using __is_vector = _IsVector;
    // backend tag can be change depending on
    // TBB availability in the environment
    using __backend_tag = _BackendType ;
};
```

Reply:
And for `__hetero_tag` too?

Reply:
And for `__parallel_forward_tag` too?

Reply:
Fixed for `__parallel_tag` and for `__parallel_forward_tag`

Reply:
sounds good

Reply:
Let's do this exercise later: looks like it's not too simple.

Reply:
Fixed.